modVM/src/org/aion/vm/VirtualMachineFactory.java

```diff
@@ -0,0 +1,131 @@
+package org.aion.vm;
+
+import org.aion.avm.core.AvmImpl;
+import org.aion.avm.core.CommonAvmFactory;
+import org.aion.fastvm.FastVirtualMachine;
+import org.aion.vm.api.interfaces.KernelInterface;
+import org.aion.vm.api.interfaces.VirtualMachine;
+
+/**
+ * A Singleton factory class that is responsible for:
```

Comment:
A few thoughts on this class...

Not a big fan of singleton pattern (for the reasons summarized in https://softwareengineering.stackexchange.com/a/40610) -- is it necessary to have that restriction?  I think it would be better to instantiate it in Aion.main and pass it to the places that need it.

Also, right now it can't be subclassed (due to `final` and the singleton pattern).  Since the vision is that people can implement their own VM if they like and use it in our code, that seems counter-intuitive.

I think 'Factory' is a bit of a misleading name since the class does work other than constructing virtual machines (it tracks their state, etc).  Would it be better to call it VirtualMachineManager or something like that?

Reply:
I've changed the name to manager.
I do think the singleton/final pattern is appropriate for this class though, because we only want one authoritative manager. If a new vm implementation is added, that's no worry, you simply add it to the supported vm's in the manager (the manager itself is not a virtual machine, so subclassing is not an issue here).
However, the one exception to this is testing, where I agree we may want to mock the vm, so I've created a Provider class that provides the caller with the manager. This is where a test can set their mock manager, in the provider. This way the actual implementation doesn't have to inherit these testing concerns